### DISCUSSION TITLE
[Usage] Fine-tuning VITS with VCTK-like dataset

### DISCUSSION THREAD

**Discussion by skol101** [espnet/espnet#3828] [Q&A] [answered]
Do I have create label files for my VCTK-like dataset? How do I do that? 
   
This the command I'm running
```
./run.sh     --stage 1     --stop-stage 5     --g2p g2p_en_no_space     --min_wav_duration 0.38     --fs 22050     --n_fft 1024     --n_shift 256     --dumpdir dump22k     --win_length null     --tts_task gan_tts     --feats_extract linear_spectrogram     --feats_normalize none     --train_config ./conf/tuning/train_multi_spk_vits.yaml 
```
The first ouput of this command is this
```
2021-11-30T19:00:25 (tts.sh:205:main) ./tts.sh --lang en --feats_type raw --fs 24000 --n_fft 2048 --n_shift 300 --win_length 1200 --token_type phn --cleaner tacotron --g2p g2p_en_no_space --train_config conf/train.yaml --inference_config conf/decode.yaml --train_set tr_no_dev --valid_set dev --test_sets dev eval1 --srctexts data/tr_no_dev/text --audio_format flac --stage 1 --stop-stage 5 --g2p g2p_en_no_space --min_wav_duration 0.38 --fs 22050 --n_fft 1024 --n_shift 256 --dumpdir dump22k --win_length null --tts_task gan_tts --feats_extract linear_spectrogram --feats_normalize none --train_config ./conf/tuning/train_multi_spk_vits.yaml
```
So sample rate from 22050 magically turns to 2400? How?

**Comment by kan-bayashi** [comment, marked as answer]
First, each recipe is made for each corpus.
So we recommend creating the recipe for your corpus at first.

> Do I have create label files for my VCTK-like dataset? How do I do that?

The label file is used to create segments file, which is used to trim the long silence at the beginning and the end of each audio (Since VCTK samples contain a long silence, it is needed to trim them for better quality).
So that is an optional, if your dataset is well organized, you do not need to prepare.
What you need to prepare is the kaldi-style directory including
- wav.scp
- text
- utt2spk
- spk2utt (can be created from utt2spk) 


> So sample rate from 22050 magically turns to 2400? How?

Sorry for the confusion. 
The command line option will be overwritten with the latter one.
So in this case, --fs 22050 will be used.

**Reply by skol101** [reply]
Could you help me here: if we have N number of my speakers, we cannot just fine tune pretrained VITS VCTK model, instead we have to train the whole joint dataset from scratch?

And do it every time we add new speakers ?

**Reply by skol101** [reply]
> What you need to prepare is the kaldi-style directory including

So I have to follow this route https://github.com/espnet/espnet/blob/079d3b25da0e2a061c2bb8452d54389e17a32118/egs2/TEMPLATE/tts1/README.md#1-data-preparation ? Do I have to lump both VCTK dataset and my dataset together as I understand finetuning to multiple new speakers doesn't really work with VITS?

**Reply by kan-bayashi** [reply]
I think it depends on what you want to do.
If you want to synthesize only the speech of newly added speaker, you can do fine-tuning.
Maybe in this case, X-vector based model is easy to fine-tune.
Just fine-tuning the model with your new dataset.
But maybe in this case, I'm not sure the quality of the original training speakers will be.

**Reply by kan-bayashi** [reply]
If you want to keep the quality of the original speakers, why don't you finetuning with original vctk dataset + your dataset? Maybe this is better than the training from scratch in terms of the training cost.

**Reply by kan-bayashi** [reply]
Related: https://github.com/espnet/espnet/issues/3705

**Reply by kan-bayashi** [reply]
If you want to finetune the model with speaker id, it is a little bit complicated.
You need to extend the speaker id dictionary to include your new speakers and modify the speaker embedding size if the number of speakers exceed the size of speaker embedding.

**Reply by kan-bayashi** [reply]
> So I have to follow this route 079d3b2/egs2/TEMPLATE/tts1/README.md#1-data-preparation ? Do I have to lump both VCTK dataset and my dataset together as I understand finetuning to multiple new speakers doesn't really work with VITS?

If you want to concatenate your dataset with vctk one, there are two options:
- Add your new speaker dataset in vctk dataset by using the same format.
- Prepare kaldi-style data directory somehow and combine it with vctk one. You can use `utils/combine_data.sh <concat_data_dir> <vctk_data_dir> <your_data_dir>`.

**Reply by skol101** [reply]
I want to have multiple new speakers to add to VCTK dataset, and use the trained model to synthensize text only for new speakers. And then to extend model to even newer speakers without having to spend at least a week again on retraining. 

>You need to extend the speaker id dictionary to include your new speakers and modify the speaker embedding size if the number of speakers exceed the size of speaker embedding.

Is that the path I should pursue?

**Reply by kan-bayashi** [reply]
> Is that the path I should pursue?

Then, I recommend trying the fine-tuning of X-vector based model.
(X-vector model is better than SID model from our experiences.)
In terms of fine-tuning, that is more easier.

**Reply by zzxiang** [reply]
> 

Sorry to jump in. I'm a bit interested in X-vector model. It seems that X-vector model means multi-speaker model, and SID model means single speaker model. Is my understanding correct?

**Reply by kan-bayashi** [reply]
No. Both are multi-speaker model. 
X-vector model uses pretrained x-vector as the auxiliary input.
SID (Speaker ID) model use one-hot speaker ID as the auxiliary input.

**Reply by zzxiang** [reply]
Thanks for explanation!

**Reply by skol101** [reply]
Looks like there's no pre-trained x-vector VITS model based on VCTK dataset, correct? https://github.com/espnet/espnet_model_zoo/blob/master/espnet_model_zoo/table.csv

Or VITS model doesn't support X-vector?

Also, GPU memory wise I've read on CoquiTTS that VITS VCTK model is very GPU memory hungry -- memory usage is increasing with number of characters in text. Is that true here as well?

**Reply by skol101** [reply]
> No. Both are multi-speaker model. X-vector model uses pretrained x-vector as the auxiliary input. SID (Speaker ID) model use one-hot speaker ID as the auxiliary input.

Does it use a random one-hot speaker? If yes,does it make sense to split VCTK into male/female sets and train separately?

**Reply by kan-bayashi** [reply]
> Or VITS model doesn't support X-vector?

You can use:
- https://github.com/espnet/espnet/tree/master/egs2/vctk/tts1#vctk_tts_train_xvector_vits_raw_phn_tacotron_espeak_ng_english_us_vits_traintotal_countave
- https://github.com/espnet/espnet/tree/master/egs2/libritts/tts1#libritts_tts_train_xvector_vits_raw_phn_tacotron_espeak_ng_english_us_vits_traintotal_countave
- https://github.com/espnet/espnet/tree/master/egs2/libritts/tts1#libritts_tts_train_xvector_vits_raw_phn_tacotron_g2p_en_no_space_traintotal_countave

> Also, GPU memory wise I've read on CoquiTTS that VITS VCTK model is very GPU memory hungry -- memory usage is increasing with number of characters in text. Is that true here as well?

We use dynamic batch making strategy so the usage of GPU memory should not depends on the dataset. In my experiments, I used V100 x 4 (each gpu requires around 12GB).

**Reply by kan-bayashi** [reply]
> Does it use a random one-hot speaker? If yes,does it make sense to split VCTK into male/female sets and train separately?

What does it mean "random hot-speaker"? We used the specific ID for each speaker.

**Reply by skol101** [reply]
Curious thing, but if lab directory isn't present for a speaker, the data_prep.sh won't proceed
```
for spk in ${spks}; do
    # check spk existence                                                                                         
    [ ! -e "${db}/lab/mono/${spk}" ] && \
        echo "${spk} does not exist." >&2 && exit 1;
```
And if I put empty p0speaker directory into /lab/mono the data_prep.sh will stop like this
```
p0speaker_397 does not have a label file. skipped.
p0speaker_398 does not have a label file. skipped.
p0speaker_399 does not have a label file. skipped.
local/data_prep.sh: line 110: data/p0speaker_train/wav.scp: No such file or directory
``` 
But you said above that LAB files are optional. Though without them new speaker can't be added to the train/dev. 
Looks like it's better to preprocess my VCTK-like dataset through this process to create LAB files https://github.com/r9y9/deepvoice3_pytorch/tree/master/vctk_preprocess

**Reply by skol101** [reply]
> We use dynamic batch making strategy so the usage of GPU memory should not depends on the dataset. In my experiments, I used >V100 x 4 (each gpu requires around 12GB).

I meant memory usage at VITS inference  -- the longer the input sentence, the higher memory consumption.

**Reply by kan-bayashi** [reply]
> But you said above that LAB files are optional. Though without them new speaker can't be added to the train/dev.

Sorry for my poor explanation. I mean `.lab` file is not need in general, but if you want to use vctk script as is, it is necessary. I made workaround script, if you do not want to prepare label file, please try it (Not tested and maybe it requires compiled kaldi).
https://gist.github.com/kan-bayashi/eceafcd35a2351f5f6bf89a1ccb956e9

> Looks like it's better to preprocess my VCTK-like dataset through this process to create LAB files r9y9/deepvoice3_pytorch@master/vctk_preprocess

 That is the best way.

**Reply by kan-bayashi** [reply]
> I meant memory usage at VITS inference -- the longer the input sentence, the higher memory consumption.

Yes, longer sentence requires larger memory but that is the same for the other models.
Actually I've never strict memory consumption in the inference.
Maybe you can quickly check via colab demo.
https://colab.research.google.com/github/espnet/notebook/blob/master/espnet2_tts_realtime_demo.ipynb

**Reply by skol101** [reply]
Discussion about VITS memory usage was here https://github.com/coqui-ai/TTS/discussions/862

**Reply by skol101** [reply]
Do you know of other ways to create VCTK labels?

I found it: https://github.com/CSTR-Edinburgh/merlin . That's all that's needed to create lab files for VCTK like dataset.

**Reply by skol101** [reply]
> First, each recipe is made for each corpus. So we recommend creating the recipe for your corpus at first.
> 
> > Do I have create label files for my VCTK-like dataset? How do I do that?
> 
> The label file is used to create segments file, which is used to trim the long silence at the beginning and the end of each audio (Since VCTK samples contain a long silence, it is needed to trim them for better quality). So that is an optional, if your dataset is well organized, you do not need to prepare. What you need to prepare is the kaldi-style directory including
> 
>     * wav.scp
> 
>     * text
> 
>     * utt2spk
> 
>     * spk2utt (can be created from utt2spk)

I have KALDI installed it's unclear how to prepare dataset in kaldi format for ESPNET

**Reply by kan-bayashi** [reply]
Please check
https://github.com/espnet/espnet/tree/master/egs2/TEMPLATE#about-kaldi-style-data-directory

**Comment by skol101** [comment]
@kan-bayashi with merlin  ./02_prepare_labels.sh 
I managed to create Full context labels, but not mono labels. 
I got all mono labels in one file mono_align.mlf.

Question: How do you generate mono labels for each file separately? 
Answer: have to comment out "rm" lines in scripts/prepare_labels_from_txt.sh so that script leaves out prompt-lab directory where full and mono labels are stored.

The whole thing is unnecessary complex, though. Why CoquiTTS can just use VCTK dataloader to load raw wav/txt files and here so many hoops are required to train on VCTK?

**Reply by kan-bayashi** [reply]
> The whole thing is unnecessary complex, though. Why CoquiTTS can just use VCTK dataloader to load raw wav/txt files and here so many hoops are required to train on VCTK?

Thank you for your patience.
We can use the vctk data as is, but it will result in the poor performance.
Careful trimming is key to build a model.
Since VCTK trimming is not good in original, so we perform the force alignment.

**Reply by skol101** [reply]
Please, explain: if I want to use x-vector to fine tune to TWO NEW (non VCTK) speakers, do I have to have complete VCTK-corpus? I mean, this model https://zenodo.org/record/5560146#.YbB5SruxXJw is already trained on the complete VCTK, is it required to run stage-1 on the full dataset?

**Reply by kan-bayashi** [reply]
> is it required to run stage-1 on the full dataset?

If you want to just load the pretrained weight, that is not needed.
Basically, we create the recipe for each corpus, so we did not design each recipe to use with the other corpora.
So, if you want to perform adaptation with your dataset, I recommend creating the recipe for you corpus.

**Reply by kan-bayashi** [reply]
> Also, this command has fs 22050 but uses full band tuning conf file which has fs setting of 44100. I guess I have to create my conf file to with my settings.

Oh sorry, this is my typo. train_config should be `./conf/tuning/train_xvector_vits.yaml`.

**Reply by skol101** [reply]
I'm runing egs2 recipe 
```
/run.sh     --train_set tr_no_dev_phn     --valid_set dev_phn     --test_sets "dev_phn eval1_phn"     --srctexts "data/tr_no_dev_phn/text"     --g2p none     --cleaner none     --stage 2     --use_xvector true     --min_wav_duration 0.38     --ngpu 3     --fs 22050     --n_fft 1024     --n_shift 256     --dumpdir dump/22k     --expdir exp/22k     --win_length null     --tts_task gan_tts     --feats_extract linear_spectrogram     --feats_normalize none     --train_config ./conf/tuning/train_multi_spk_vits.yaml     --inference_model train.total_count.ave.pth
```
, but got this in my make_mfcc logs
```
# extract-segments scp,p:dump/22k/mfcc/tr_no_dev_phn/wav.scp dump/22k/mfcc/tr_no_dev_phn/log/segments.19 ark:- | compute-mfcc-feats --write-utt2dur=ark,t:dump/22k/mfcc/tr_no_dev_phn/log/utt2dur.19 --verbose=2 --config=conf/mfcc.conf ark:- ark:- | copy-feats --compress=true --write-num-frames=ark,t:dump/22k/mfcc/tr_no_dev_phn/log/utt2num_frames.19 ark:- ark,scp:/home/sk/work/espnet/egs2/vctk/vits2speak/dump/22k/mfcc/tr_no_dev_phn/data/raw_mfcc_tr_no_dev_phn.19.ark,/home/sk/work/espnet/egs2/vctk/vits2speak/dump/22k/mfcc/tr_no_dev_phn/data/raw_mfcc_tr_no_dev_phn.19.scp 
# Started at Wed Dec  8 13:16:08 EET 2021
#
bash: line 1: extract-segments: command not found
bash: line 1: compute-mfcc-feats: command not found
bash: line 1: copy-feats: command not found
# Accounting: time=0 threads=1
# Ended (code 127) at Wed Dec  8 13:16:08 EET 2021, elapsed time 0 seconds
```
Do I have still to install KALDI?

Also, there's no train_xvector_vits.yaml in https://github.com/espnet/espnet/tree/master/egs2/vctk/tts1/conf/tuning

**Reply by kan-bayashi** [reply]
Sorry, if you want to run x-vector based model, you need to compile kaldi.
https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/tts1/README.md#2-wav-dump--embedding-preparation

**Reply by skol101** [reply]
Installed KALDI, now run.sh stops here.  
I've set --sample-frequency=22050 in mfcc.conf. All of my dataset's wav files are 22050 Hz. 
```
ERROR (compute-mfcc-feats[5.5.0~1-5cd9]:ComputeFeatures():feat/feature-common-inl.h:47) Waveform and config sample Frequency mismatch: 16000 .vs 22050 (use --allow-upsample=true option to allow  upsampling the waveform).
```

**Reply by kan-bayashi** [reply]
You do not need to modify mfcc.conf since x-vector is extracted with 16k.
And for mfcc extraction, we perform downsampling automatically.

**Reply by skol101** [reply]
Cheers!
Now going one step back to Merlin.  
How did you generate mono state aligned VCTK label files? With merlin I was able to generate lab files in label_state_align which are correctly aligned to audio, but those are 'full' files, yet I don't have 'mono' state aligned files.

**Reply by kan-bayashi** [reply]
I followed r9y9/deepvoice3_pytorch implemenation.
https://github.com/kan-bayashi/VCTKCorpusFullContextLabel

Again, the label files are only used for trimming the beginning and the end of silence.
https://github.com/espnet/espnet/blob/799e24d0fcbb97ee8bafa9f4e5a91da25881117b/egs2/vctk/tts1/local/data_prep.sh#L83-L106
Here, just making segments file which represents start and end time in seconds.
So you can use full context label if you modify this part to deal with full context label.

This one may help you to create segments file from full context label.
https://github.com/espnet/espnet/blob/799e24d0fcbb97ee8bafa9f4e5a91da25881117b/egs/jsut/tts1/local/prep_segments.py

**Reply by skol101** [reply]
The problem I have is that full labels are correctly aligned, while mono labels are not aligned. And because of that KALDI incorrectly segments file: last entrance in label mono is showing that the utterance 'pau' occurs beyong the actual audio end.

**Comment by skol101** [comment]
Ok, I used your gist from here https://gist.github.com/kan-bayashi/eceafcd35a2351f5f6bf89a1ccb956e9 to skip the LAB file creation process.

I also copied train_xvector_vits from LibriTTS recipe https://github.com/espnet/espnet/blob/9823e7843545d5a30b19e670f2028259f1d0ed03/egs2/libritts/tts1/conf/tuning/train_xvector_vits.yaml into my VCTK folder.

Let's see how it trains.

**Comment by skol101** [comment]
After 75 epochs training results. Does it make sense to proceed further?

![discriminator_fake_loss](https://user-images.githubusercontent.com/53978091/145353522-70940118-a3a6-445a-99e1-e55209da2a25.png)
![discriminator_loss](https://user-images.githubusercontent.com/53978091/145353531-43d6fab7-4058-43c0-a765-2b2e0cb58dc6.png)
![discriminator_real_loss](https://user-images.githubusercontent.com/53978091/145353541-663d3263-6ee8-45e3-a952-05dadb31199a.png)
![generator_adv_loss](https://user-images.githubusercontent.com/53978091/145353554-8c07af83-3001-4138-bfb3-bb523c8189eb.png)
![generator_dur_loss](https://user-images.githubusercontent.com/53978091/145353563-9d52a875-115a-431e-b233-32d5523ac65c.png)
![generator_feat_match_loss](https://user-images.githubusercontent.com/53978091/145353567-df08833d-fd6e-4dd0-9c06-6f6f47aa5b78.png)
![generator_kl_loss](https://user-images.githubusercontent.com/53978091/145353573-c12511db-4b55-4a49-8198-965b44318235.png)
![generator_loss](https://user-images.githubusercontent.com/53978091/145353576-2f189de3-1dc0-403c-b488-67980a88d92c.png)
![generator_mel_loss](https://user-images.githubusercontent.com/53978091/145353577-46ddc635-a202-4cc7-9f68-22f6f409a34e.png)

**Reply by kan-bayashi** [reply]
It seems fine. You can compare the graph included in my pretrained model.
https://zenodo.org/record/5560146
And I recommend generating the intermediate results to check the quality.
```
# e.g., 
./run.sh \
  --stage 7 \
  --train_set tr_no_dev_phn \
  --valid_set dev_phn \
  --test_sets "dev_phn eval1_phn" \
  --srctexts "data/tr_no_dev_phn/text" \
  --g2p none \
  --cleaner none \
  --use_xvector true \
  --min_wav_duration 0.38 \
  --ngpu 3 \
  --fs 22050 \
  --n_fft 1024 \
  --n_shift 256 \
  --dumpdir dump/22k \
  --expdir exp/22k \
  --win_length null \
  --tts_task gan_tts \
  --feats_extract linear_spectrogram \
  --feats_normalize none \
  --train_config ./conf/tuning/train_xvector_vits.yaml \
  --inference_model 75epoch.pth
```

**Reply by skol101** [reply]
Thank you! Generated voice isn't bad at all, though where do I set actual sentences, because what I hear from the generated wavs is "Andy space W space WUE space ROID space NWN", where capital letters are pronounced as abbreviations.

Regarding params: does it matter if I had g2p param as 'g2p_en_no_space' and cleaner as 'tacotron'?

**Reply by kan-bayashi** [reply]
> Regarding params: does it matter if I had g2p param as 'g2p_en_no_space' and cleaner as 'tacotron'?

In our experiments, espeak based G2P is more accurate than G2P_en although G2P_en is much faster than espeak.
And VITS seems more sensitive to the G2P errors.
Please check the samples of "Analysis of wrong G2P results" in HP.
https://espnet.github.io/icassp2022-tts/

**Reply by skol101** [reply]
> It seems fine. You can compare the graph included in my pretrained model. https://zenodo.org/record/5560146 And I recommend generating the intermediate results to check the quality.


Oh I should have fine tuned this model , though, it's not available on Hugging Face. Would you check please
https://huggingface.co/api/models/kan-bayashi/vctk_tts_train_xvector_vits_raw_phn_tacotron_espeak_ng_english_us_vits_train.total_count.ave/revision/main

**Reply by kan-bayashi** [reply]
Please download from zenodo.
https://zenodo.org/record/5560146/files/tts_train_xvector_vits_raw_phn_tacotron_espeak_ng_english_us_vits_train.total_count.ave.zip?download=1

**Reply by skol101** [reply]
When running stage 1 as per https://github.com/espnet/espnet/blob/9823e7843545d5a30b19e670f2028259f1d0ed03/egs2/jvs/tts1/README.md#1-run-the-recipe-until-stage-5-with-2205khz-setup
```
2021-12-09T12:02:23,001 (tts.sh:500:main) Stage 4: Generate token_list from data/tr_no_dev/text
/home/sk/anaconda3/envs/espnet/bin/python3 /home/sk/work/espnet/espnet2/bin/tokenize_text.py --token_type phn -f 2- --input dump/22k/raw/srctexts --output dump/22k/token_list/phn_tacotron_espeak_ng_english_us_vits/tokens.txt --non_linguistic_symbols none --cleaner tacotron --g2p espeak_ng_english_us_vits --write_vocabulary true --add_symbol '<blank>:0' --add_symbol '<unk>:1' --add_symbol '<sos/eos>:-1'
2021-12-09 11:02:23,023 (words_mismatch:83) WARNING: words count mismatch on 100.0% of the lines (1/1)
2021-12-09 11:02:23,991 (words_mismatch:83) WARNING: words count mismatch on 200.0% of the lines (2/1)
2021-12-09 11:02:24,476 (words_mismatch:83) WARNING: words count mismatch on 100.0% of the lines (1/1)
2021-12-09 11:02:29,357 (words_mismatch:83) WARNING: words count mismatch on 100.0% of the lines (1/1)
2021-12-09 11:02:30,826 (words_mismatch:83) WARNING: words count mismatch on 100.0% of the lines (1/1)
2021-12-09 11:02:35,278 (words_mismatch:83) WARNING: words count mismatch on 100.0% of the lines (1/1)
```
Should I run stage 4 manually with an option --words-mismatch for a phonemizer?

**Reply by skol101** [reply]
This guide is very helpful in fine tuning https://github.com/espnet/espnet/blob/9823e7843545d5a30b19e670f2028259f1d0ed03/egs2/jvs/tts1/README.md
Cheers!

**Reply by skol101** [reply]
But still stops at random point during first epoch
```
CUDA_VISIBLE_DEVICES="0,2"  ./run.sh     --stage 6     --g2p espeak_ng_english_us_vits     --min_wav_duration 0.38 --ngpu 2     --fs 22050     --n_fft 1024     --n_shift 256     --dumpdir dump/22k     --win_length null     --tts_task gan_tts     --feats_extract linear_spectrogram     --feats_normalize none     --train_config ./conf/tuning/train_xvector_vits.yaml     --train_args "--init_param downloads/tts-train-xvec-vits/exp/tts_train_xvector_vits_raw_phn_tacotron_espeak_ng_english_us_vits/train.total_count.ave_5best.pth:tts:tts"     --tag finetune_vits_raw_phn_espeak_ng_english_us --use_xvector true
```

```
[bigx:0/2] 2021-12-09 12:56:06,908 (gan_trainer:305) INFO: 1epoch:train:351-400batch: iter_time=2.979, discriminator_forward_time=0.186, discriminator_loss=2.357, discriminator_real_loss=1.234, discriminator_fake_loss=1.123, discriminator_backward_time=0.152, discriminator_optim_step_time=0.009, optim1_lr0=2.000e-04, discriminator_train_time=0.377, generator_forward_time=0.066, generator_loss=45.743, generator_mel_loss=24.644, generator_kl_loss=3.014, generator_dur_loss=2.160, generator_adv_loss=2.629, generator_feat_match_loss=13.296, generator_backward_time=0.278, generator_optim_step_time=0.058, optim0_lr0=2.000e-04, generator_train_time=0.447, train_time=3.810
[bigx:0/2] 2021-12-09 13:00:11,481 (gan_trainer:305) INFO: 1epoch:train:401-450batch: iter_time=4.064, discriminator_forward_time=0.183, discriminator_loss=2.334, discriminator_real_loss=1.241, discriminator_fake_loss=1.093, discriminator_backward_time=0.154, discriminator_optim_step_time=0.009, optim1_lr0=2.000e-04, discriminator_train_time=0.375, generator_forward_time=0.065, generator_loss=45.610, generator_mel_loss=24.361, generator_kl_loss=2.955, generator_dur_loss=2.139, generator_adv_loss=2.642, generator_feat_match_loss=13.512, generator_backward_time=0.278, generator_optim_step_time=0.057, optim0_lr0=2.000e-04, generator_train_time=0.443, train_time=4.891
[bigx:0/2] 2021-12-09 13:04:55,104 (gan_trainer:305) INFO: 1epoch:train:451-500batch: iter_time=4.826, discriminator_forward_time=0.187, discriminator_loss=2.328, discriminator_real_loss=1.215, discriminator_fake_loss=1.114, discriminator_backward_time=0.157, discriminator_optim_step_time=0.009, optim1_lr0=2.000e-04, discriminator_train_time=0.384, generator_forward_time=0.066, generator_loss=45.507, generator_mel_loss=24.387, generator_kl_loss=2.877, generator_dur_loss=2.141, generator_adv_loss=2.634, generator_feat_match_loss=13.468, generator_backward_time=0.283, generator_optim_step_time=0.060, optim0_lr0=2.000e-04, generator_train_time=0.453, train_time=5.672

python3: src/libespeak-ng/fifo.c:90: fifo_init: Assertion `0' failed.
```
or with this error
```
libespeak-ng.so.1.1.51: failed to map segment from shared object
```

**Reply by kan-bayashi** [reply]
> libespeak-ng.so.1.1.51: failed to map segment from shared object

It seems new version of phonemizer is not stable.
I recommend performing the conversion as a preprocessing instead of the on-the-fly g2p.

**Reply by skol101** [reply]
@kan-bayashi , I'm running inference notebook, but it always generates one sound "ehh", even though dev wav are full sentences. Please check this collab.
https://colab.research.google.com/drive/1Co-EwTPBY59RF1nu-vFXm5oSqaN42Ih5?usp=sharing

**Reply by kan-bayashi** [reply]
Since you train the model with `g2p=none` and `cleaner=none`, the model assumes the phoneme as the inputs. If you want to deploy, after training, please edit exp/your_model_dir/config.yaml to modify `g2p: espeak_ng_english_us_vits` and `cleaner: tacotron`. Then the model accepts the standard English sentence.

**Reply by skol101** [reply]
Thank you! I'm quite satisfied with fine tuning result on my custom dataset. I wonder if I could achieve better result if I were to jointly train FS2/FastPitch and HiFiGan (or Stylegan) vocoder?

**Reply by kan-bayashi** [reply]
I'm glad to hear that.
From my experiments, in the multi-speaker model, VITS is definitely better than FS2.
In the case of single speaker model, jointly trained FS2 is comparable quality with better controllability.

**Reply by skol101** [reply]
Is it possible to train VITS with MB-MelGan as a vocoder? Just for comparison sake.

**Reply by kan-bayashi** [reply]
It can be but you need to modify the code by yourself.
https://github.com/espnet/espnet/blob/9d08012fa8f4a20053d19d82316b4ae093ad088a/espnet2/gan_tts/vits/generator.py#L194
You can replace the decoder with MB-melgan.

**Reply by skol101** [reply]
Cheers! This would require training VITS VCTK from the scratch, right?

**Reply by kan-bayashi** [reply]
Right.

**Reply by skol101** [reply]
Should config.yaml file be changed for inference? For instance, would use_amp: true have any effect on inference speed?

**Reply by kan-bayashi** [reply]
amp affects only the training.
Instead, you can try `dtype` option for the inference (I'm not sure working well).
https://github.com/espnet/espnet/blob/bf523b70cae8300da004b41ec6a0d1b57c7ae8bb/espnet2/bin/tts_inference.py#L604

**Comment by skol101** [comment]
I wonder, if I add more voices to the dataset, should I finetune from the ESPNet trained model (1mln steps) or from my own finetuned 'best of 10' model?

**Reply by skol101** [reply]
Can finetuned VITS model be further 'finetuned' on a new voice? @kan-bayashi

**Reply by kan-bayashi** [reply]
I think it can be finetuned. I’ve never tried with vits but played with hifigan and it worked well.
